examples/crash_recovery/src/main.rs

```diff
@@ -0,0 +1,149 @@
+use log::{debug, error, info};
+use network::Network;
+use std::{collections::HashMap, fs, fs::File, io, io::Write, path::Path};
+
+/// Example node producing linear order.
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct Args {
+    /// Index of the node
+    #[clap(long, value_parser)]
+    id: usize,
+
+    /// Ports
+    #[clap(long, value_parser, value_delimiter = ',')]
+    ports: Vec<usize>,
+
+    /// Number of items to be ordered
+    #[clap(long, value_parser)]
+    n_ordered: u32,
+
+    /// Number of items to be created
+    #[clap(long, value_parser)]
+    n_created: u32,
+
+    /// Number of the first created item
+    #[clap(long, value_parser)]
+    n_starting: u32,
+
+    /// Should the node crash after finalizing its items
+    #[clap(long, value_parser)]
+    crash: bool,
+}
+
+fn create_backup(node_id: usize) -> Result<(File, io::Cursor<Vec<u8>>), io::Error> {
+    let stash_path = Path::new("./aleph-bft-examples-crash-recovery-backup");
+    fs::create_dir_all(&stash_path)?;
+    let file_path = stash_path.join(format!("{}.units", node_id));
+    let _ = fs::OpenOptions::new()
+        .create(true)
+        .append(true)
+        .open(file_path.clone());
+    let loader = io::Cursor::new(fs::read(file_path.clone())?);
+    let saver = fs::OpenOptions::new()
+        .write(true)
+        .append(true)
+        .open(file_path)?;
+    Ok((saver, loader))
+}
+
+#[tokio::main]
+async fn main() {
+    env_logger::builder()
+        .format(|buf, record| {
+            writeln!(
+                buf,
+                "{} {}: {}",
+                record.level(),
+                Local::now().format("%Y-%m-%d %H:%M:%S%.3f"),
+                record.args()
+            )
+        })
+        .filter(None, log::LevelFilter::Debug)
+        .init();
+
+    let args = Args::parse();
+
+    info!("Getting network up.");
+    let network = Network::new(args.id, &args.ports).await.unwrap();
+    let n_members = args.ports.len();
+    let data_provider = DataProvider::new(args.id.into(), args.n_starting, args.n_created);
+    let (finalization_handler, mut finalized_rx) = FinalizationHandler::new();
+    let (backup_saver, backup_loader) =
+        create_backup(args.id).expect("Error setting up unit saving");
+    let local_io = aleph_bft::LocalIO::new(
+        data_provider,
+        finalization_handler,
+        backup_saver,
+        backup_loader,
+    );
+
+    let (close_member, exit) = oneshot::channel();
+    let member_handle = tokio::spawn(async move {
+        let keychain = Keychain::new(n_members.into(), args.id.into());
+        let config = aleph_bft::default_config(n_members.into(), args.id.into(), 0);
+        run_session(config, local_io, network, keychain, Spawner {}, exit).await
+    });
+
+    let mut count_finalized: HashMap<NodeIndex, u32> =
+        (0..args.ports.len()).map(|c| (c.into(), 0)).collect();
+    let show_finalized = |cf: &HashMap<NodeIndex, u32>| -> Vec<u32> {
```

Comment:
I think @DamianStraszak wanted to have this sorted by `NodeIndex` in another PR, and also suggested extracting the actual number from `NodeIndex` before logging it

Reply:
and why not an anonymous function?

Reply:
Yeah, default print of a hashmap does not preserve the order of keys, which is then annoying to read. Anonymous function here sounds OK to me, although `show_finalized` is not a perfect name for it -- maybe `finalized_counts` or some such?

Reply:
renamed

Reply:
I see a point of using closures if you have some local variable in this context that you would like to move, if we pass this function somewhere or in case of iterators and so on. But if the function uses only arguments and is only called here, this seems a little unnatural to me. It is purely stylistic preference and it seems to me, like in our code base we almost never use closures in such way.